### PR TITLE
Attack command: Fix ground Area Attack target cycling

### LINKED ISSUE #8772: Attack command: Area attack for ground always shoots one spot
### Description

Area attack for ground units does not work as intended. When an artillery unit is ordered to area attack empty ground, it chooses one random point inside the circle and keeps attacking that same point. It should choose new random points throughout the area.

### Steps to reproduce

1. Start a game and create a ground artillery unit with area attack, for example an `armart`.
2. Give it an area-attack command on empty ground.
3. Observe where successive shots land.

### Expected behavior

The unit attacks different random points inside the selected circle, as described by the command tooltip: “attack an area randomly”.

### Actual behavior

The unit randomly chooses one point and continues attacking only that point.

### Cause

Ground units use the Lua `CMD_AREA_ATTACK_GROUND` implementation in `luarules/gadgets/unit_areaattack.lua`, rather than the engine-native `CMD_AREA_ATTACK` used by aircraft.

The gadget chooses a random position and inserts an ordinary ground `CMD_ATTACK` at the front of the unit's command queue. An ordinary artillery ground-attack order does not finish after firing, so the custom area-attack command behind it is never processed again and cannot choose another position.

The focused game-side fix is tracked in #8773.

### Video

Before

https://github.com/user-attachments/assets/4a19f803-831e-4e8b-a6c6-ccfc0bc2ee50

After

https://github.com/user-attachments/assets/d90fdcac-b56c-4421-a05f-175c32ca43f1

### LINKED ISSUE #8774: Attack Command: Queued ground attacks do not advance past the first position
### Description

When multiple artillery ground-attack commands are queued, the unit keeps firing at the first position forever and never advances through the queue.

### Steps to reproduce

1. Start a game and create a ground artillery unit, for example an `armart` or `corwolv`.
2. Hold Shift and queue ground-attack commands at several different positions.
3. Observe the unit after it fires at the first position.

### Expected behavior

The unit fires one complete salvo at each queued position and then advances to the next command. The final ground-attack command may remain persistent because no further command follows it.

### Actual behavior

The first ground-attack command never finishes. The unit continues firing at the first position and does not process the remaining queued attacks.

### Video

Notice the Wolverine keeps shooting at the same point:

https://github.com/user-attachments/assets/01cb2af2-27b7-4366-bd45-b56ff0706e49

### Proposed implementations

- #8978 adds a configurable to advance the ground attack after a per unit specified salvo count.
- #8933 is a simpler solution that depends on a proposed new engine event in beyond-all-reason/RecoilEngine#3286.

### PR DESCRIPTION
**Pull request by eun-ice**
## Why

Ground Area Attack is currently stuck on its first randomly selected position. The Lua implementation inserts an ordinary ground `CMD_ATTACK` ahead of the persistent area command, but ordinary artillery ground attacks do not finish on their own. The area command therefore never gets another turn to choose a new position.

This PR fixes only the generated Area Attack command lifecycle. It does not change the behavior of player-issued or queued ground-attack commands.

Closes #8772.

Queued ground-attack behavior remains a separate design question in #8774, with the current prototype preserved in draft PR #8978.

## What

- Mark the point attacks generated by Area Attack as internal commands.
- Watch the designated Area Attack weapon and count all projectiles in one complete salvo.
- Finish only the generated point attack after that salvo so the persistent area command can choose another random position.
- Defer completion until the next game frame to avoid invalidating active engine weapon-command state.
- If another command moves ahead during that deferred frame, remove the generated attack by its exact command tag.
- Leave player-issued ground attacks, unit definitions, Repeat configuration, and ordinary command queues unchanged.

## Testing

- `stylua --check` on the gadget and focused spec.
- `luac5.1 -p` on the gadget and focused spec.
- Focused lifecycle spec: 4 successes, 0 failures, 0 errors.
- Full Lua unit suite: 76 successes, 0 failures, 0 errors.

The focused coverage verifies:

- advancement only after a complete multi-projectile salvo;
- unrelated weapon projectiles are ignored;
- generated point attacks are marked internal;
- deferred cleanup removes only the generated command by tag if another command moves ahead.

### Before

https://github.com/user-attachments/assets/4a19f803-831e-4e8b-a6c6-ccfc0bc2ee50

### After

https://github.com/user-attachments/assets/d90fdcac-b56c-4421-a05f-175c32ca43f1

## AI / LLM usage statement

OpenAI Codex was used to assist with implementation, test development, review, and splitting the original gameplay behavior change into a separate draft PR. I reviewed the resulting code and verified it with the checks listed above.

### CODE REVIEW

**Comment by TheChronographer** [comment]
I like it for area attack. But I don't know if I like it for queued attacks. What if the thing you shoot at doesn't die in one shot?

At the moment a user can queue a few attack commands, and wait for the explosion or wreckage and hit 'next' to go to the next attack or move order etc. I.e, select 5 wolverine a+right click drag a line of fire over enemy position - > queue multiple of those, and hit N when you want the line to advance. 

This would remove that in favor of a user having to queue multiple attacks in the same place, (which can cause them to be cancelled instead?). I've seen this 'volley fire' behavior implemented as a custom widget/command and I like it, I just don't know if it should be the only attack command.

**Comment by eun-ice** [comment]
@TheChronographer : Thank you for chiming in and your valuable opinion. 

I see your point, which is precisely why I split this PR into two commits, so we can have this discussion and potentially decide to only merge one.

You are absolutely right, with this change very advanced players lose "Queue A B C and than advance with N", they have to A+click the next target instead of pressing "N". 

However, those players gain the ability to "Shoot A, B, C on repeat", which was not possible before and seems like a common use case.

For new players I think it's not understandable why the queue would not advance and the current behavior feels like a bug.

If you want I can split this PR up in two so we can at least merge the area command fix until we have consensus on the command queue advance.

**Comment by TheChronographer** [comment]
Yeah, the area attack is just a flat upgrade, that needs to change. 

> However, those players gain the ability to "Shoot A, B, C on repeat", which was not possible before and seems like a common use case.

That will be useful, even if it is less efficient that 'shoot A a few times, shoot B a few times etc.'

> For new players I think it's not understandable why the queue would not advance and the current behavior feels like a bug.

That is true. But after the change can a player say 'shoot here 3 times, then shoot here 3 times, then move on'? Or will that expose a new bug. 

> If you want I can split this PR up in two so we can at least merge the area command fix until we have consensus on the command queue advance.

Maybe wait to see what someone else says. There are other options than making this (new behaviour ) the only behavior. It could be a seperate command (gamecmd.volleyattack), it could be an attack command with a modifier (like ctrl), it could be limited to some units (as command fire currently does the same for Thor, and the engine does poorly for bombers) etc.

Speaking of, does this fix bombers, which occasionally do not drop a complete attack command?

**Comment by eun-ice** [comment]
> 
> That is true. But after the change can a player say 'shoot here 3 times, then shoot here 3 times, then move on'? Or will that expose a new bug. 


That should work (TM), will provide additional testing and video tomorrow. 

However, I am not sure if repeat commands will work because this lua workaround does cmd.remove. Will provide additional testing and video. 



> > If you want I can split this PR up in two so we can at least merge the area command fix until we have consensus on the command queue advance.
> 
> 
> 
> Maybe wait to see what someone else says. There are other options than making this (new behaviour ) the only behavior. It could be a seperate command (gamecmd.volleyattack), it could be an attack command with a modifier (like ctrl), it could be limited to some units (as command fire currently does the same for Thor, and the engine does poorly for bombers) etc.


Both commits depend on each other which is why I have them in one PR. 

But if we can merge the area one quickly, then its maybe worth a separate PR. 



> 
> 
> 
> Speaking of, does this fix bombers, which occasionally do not drop a complete attack command? 

Not sure what issues you are exactly talking about.

I hope that the biggest issues are fixed by my other PR #8770 and #8771 and that #8793 will fix the rest by removing the command limiter altogether. 

If you have an issue number for me I would love to investigate more deeply whether that's one of the issues I already addressed or something different

**Comment by TheChronographer** [comment]
This issue is reporting the bomber issue, though it's not the most descriptive: https://github.com/beyond-all-reason/Beyond-All-Reason/issues/3609

**Comment by eun-ice** [comment]
> This issue is reporting the bomber issue, though it's not the most descriptive: #3609

So i I understand correctly, for bombers (their weapon has noAutoTarget true)  the engine will proceed to the next attack command once the salvo is finished. If the bombers did not complete a full salvo because of terrain or other restrictions, the engine will not move to the next command, causing #3609 

For non-bombers, the engine does not have area ground attacks, instead they are simulated in LUA, causing all the complications we have here.

A proper engine fix would be simple, but depend on [RecoilEngine#1032](https://github.com/beyond-all-reason/RecoilEngine/issues/1032). We could add a weapon def finishGroundAttackAfterSalvo and do this in engine.

**Comment by eun-ice** [comment]
> > That is true. But after the change can a player say 'shoot here 3 times, then shoot here 3 times, then move on'? Or will that expose a new bug.
> 
> That should work (TM), will provide additional testing and video tomorrow.

Sadly this is not possible: The UI won't let you queue two attacks to the same position, as queuing the same attack will remove the previous command.


> 
> However, I am not sure if repeat commands will work because this lua workaround does cmd.remove. Will provide additional testing and video.

Repeat is fixed now, I updated this PR and uploaded a new video in the description

**Comment by TheChronographer** [comment]
> bombers (their weapon has noAutoTarget true) the engine will proceed to the next attack command once the salvo is finished. If the bombers did not complete a full salvo because of terrain or other restrictions, the engine will not move to the next command.

Well, except they do finish the salvo. They drop all the bombs in the burst correctly, just the engine doesn't get the signal. I havnt looked into it deeply.

File: luarules/gadgets/unit_areaattack.lua [hunk 1]
```diff
@@ -47,6 +57,19 @@ if gadgetHandler:IsSyncedCode() then
...
+		-- active weapon-command state, so defer it until the next game frame.
+		for unitID, attack in pairs(finishedAttacks) do
+			finishedAttacks[unitID] = nil
+			local states = Spring.GetUnitStates(unitID)
+			if states and states["repeat"] and attack.repeatable then
+				-- CMD.REMOVE deliberately disables the engine's normal repeat handling.
+				-- Reproduce FinishCommand's behavior by appending a copy first.
+				giveInsertOrderToUnit(unitID, CMD_ATTACK, attack.params, attack.options, -1, CMD.OPT_ALT)
+			end
+			Spring.GiveOrderToUnit(unitID, CMD.REMOVE, { attack.commandTag }, 0)
```
Comment (sprunk) [inline luarules/gadgets/unit_areaattack.lua]: if the inserted subcommands have the `CMD.OPT_INTERNAL` option, `Spring.FinishCommand` won't repeat them

File: luarules/gadgets/unit_areaattack.lua [hunk 2]
```diff
@@ -108,15 +141,74 @@ if gadgetHandler:IsSyncedCode() then
...
+				params = currentCommand.params,
+				options = currentCommand.options,
+				repeatable = generatedAttackTags[ownerID] ~= currentCommand.tag,
+			}
+			activeAttacks[ownerID] = attack
+		elseif attack.weaponDefID ~= weaponDefID then
+			return
+		end
+
+		attack.projectilesLeft = attack.projectilesLeft - 1
```
Comment (sprunk) [inline luarules/gadgets/unit_areaattack.lua]: this looks like it would benefit if the engine exposed to lua the EndBurst event that exists for COB

Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Agreed. A synced EndBurst call-in would be better than counting ProjectileCreated events.

I’m keeping this PR game-only so it works with current engines. BAR artillery currently has one projectile per shot, including the Legion LRPC three-shot burst, so the existing counting should work (TM) for now.

Thinking about a separate PR for the EndBurst.

File: luarules/gadgets/unit_areaattack.lua (hunk 1 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Good point, addressed in c997a23d18. 

Generated area-attack subcommands now use CMD.OPT_INTERNAL, and the normal deferred path calls Spring.UnitFinishCommand, leaving Repeat and UnitCmdDone handling to the engine.

I retained manual requeue plus exact-tag removal only as a defensive fallback when another command moves ahead during the deferred frame. This prevents accidentally finishing the newly inserted current command. 

Both paths now have vibe coded test coverage. Will perform more manual testing now, I thought of more edge cases.

File: luarules/gadgets/unit_areaattack.lua (hunk 2 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Oh well, I can break it. If I queue an attack out of range or a movement command before with space+click, the unit can end up in a state  where the weapon is not shooting, the unit is not moving and nothing is advancing.

Working on that now
image

Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: this was again the issue I tried to address in https://github.com/beyond-all-reason/RecoilEngine/pull/3241

Only this time it happened with a ground attack

https://github.com/user-attachments/assets/148923cc-1c57-429d-9735-1d3d987a2de5

So unrelated to this issue.

**Review by sprunk — APPROVED** [review]
Looks like the correct general 'shape' for an implementation of cycling.

I haven't looked at whether the implementation details are correct, whether it follows other BAR code conventions, or whether cycling is what BAR game design direction wants.

**Comment by eun-ice** [comment]
> 
> 
> > bombers (their weapon has noAutoTarget true) the engine will proceed to the next attack command once the salvo is finished. If the bombers did not complete a full salvo because of terrain or other restrictions, the engine will not move to the next command.
> 
> 
> 
> Well, except they do finish the salvo. They drop all the bombs in the burst correctly, just the engine doesn't get the signal. I havnt looked into it deeply. 

That is not related. I opened a separate engine PR to fix bomber behavior.

**Comment by eun-ice** [comment]
> Looks like the correct general 'shape' for an implementation of cycling.
> 
> I haven't looked at whether the implementation details are correct, whether it follows other BAR code conventions, or whether cycling is what BAR game design direction wants.

Thank you for the review. 

I still have to figure out who/what "BAR game design direction" is, let alone what they want 🤣

To me this is an obvious bug fix.

**Comment by WatchTheFort** [comment]
@eun-ice Please remember to disclose any AI usage in the PR description, as per our AI usage policy.

**Comment by eun-ice** [comment]
> @eun-ice Please remember to disclose any AI usage in the PR description, as per our AI usage policy.

sorry, added. Don't know why I missed that.

**Comment by eun-ice** [comment]
I updated this PR to have it hopefully be accepted in an easier way.

Now it is split into three parts

1) fix area attack choosing random target inside the circle after each salvo - current game shoots at one point forever as fix for https://github.com/beyond-all-reason/Beyond-All-Reason/issues/8772
2) allow for an optional advancing of ground attack commands  if there is a command queued after that as fix for https://github.com/beyond-all-reason/Beyond-All-Reason/issues/8774
3) Move the configuration for the actual gameplay changes in a 3rd commit that explicitly enables repeat and ground attack advancing for a set of units where I think that is useful.

**Comment by eun-ice** [comment]
I removed the controversial queue advancement fix from this PR, so this only fixes the ground area attack choosing a new random position each shot.

File: luarules/gadgets/unit_areaattack.lua [hunk 3]
```diff
@@ -30,9 +32,17 @@ if gadgetHandler:IsSyncedCode() then
...
+	local areaAttackWeapons = {}
+	local areaAttackWeaponByUnitDef = {}
 	for unitDefID, unitDef in pairs(UnitDefs) do
 		if #unitDef.weapons > 0 and unitDef.customParams.canareaattack then
-			canAreaAttack[unitDefID] = WeaponDefs[unitDef.weapons[1].weaponDef].range
+			local weaponDefID = unitDef.weapons[1].weaponDef
+			local weaponDef = WeaponDefs[weaponDefID]
+			if weaponDef then
+				canAreaAttack[unitDefID] = weaponDef.range
+				areaAttackWeapons[weaponDefID] = weaponDef.salvoSize
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: An alias should identify its contents (that's the point), e.g. areaWeaponRange, areaWeaponSalvo.

File: luarules/gadgets/unit_areaattack.lua [hunk 4]
```diff
@@ -47,17 +57,35 @@ if gadgetHandler:IsSyncedCode() then
 	}
 
 	function gadget:GameFrame(f)
+		-- Removing a command from ProjectileCreated can invalidate the engine's
+		-- active weapon-command state, so defer it until the next game frame.
+		for unitID, attack in pairs(finishedAttacks) do
+			finishedAttacks[unitID] = nil
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: If the effect's delay is purely due to escaping its callin context, then remove the command in GameFramePost.

File: luarules/gadgets/unit_areaattack.lua [hunk 5]
```diff
@@ -47,17 +57,35 @@ if gadgetHandler:IsSyncedCode() then
...
+			local commandID, _, commandTag = Spring.GetUnitCurrentCommand(unitID)
+			if commandID == CMD_ATTACK and commandTag == attack.commandTag then
+				-- Preserve the engine's normal Repeat and UnitCmdDone handling. Generated
+				-- area-attack shots are internal, so the engine will not repeat them.
+				Spring.UnitFinishCommand(unitID)
+			else
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: Not really the appropriate place to put this behavior description. It seems like we are just missing these notes in LuaSyncedCtrl.cpp.lua

File: luarules/gadgets/unit_areaattack.lua [hunk 6]
```diff
@@ -108,15 +142,71 @@ if gadgetHandler:IsSyncedCode() then
 		return false
 	end
 
+	function gadget:ProjectileCreated(projectileID, ownerID, weaponDefID)
+		local salvoSize = areaAttackWeapons[weaponDefID]
+		local unitDefID = Spring.GetUnitDefID(ownerID)
+		if not salvoSize or areaAttackWeaponByUnitDef[unitDefID] ~= weaponDefID then
+			return
+		end
+
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua:158]: ProjectileCreated is a loud callin, at the moment. We runs many membership checks (against callin x projectile) with no prefilter, e.g. for unwanted weaponDefIDs.

This can be written better to emphasize exiting early, then. If the plan is to extend Area Attack (Ground) to many, many units, though, then the :PC path should be less hot.

File: luarules/gadgets/unit_areaattack.lua [hunk 7]
```diff
@@ -108,15 +142,71 @@ if gadgetHandler:IsSyncedCode() then
...
+		local commands = Spring.GetUnitCommands(ownerID, 2)
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: Generally, prefer to retrieve the flat/multivalue return results from the engine, rather than materializing tables. OK if the code is better off with a table preformed & able to be reused; this could qualify.

File: luarules/gadgets/unit_areaattack.lua [hunk 8]
```diff
@@ -108,15 +142,71 @@ if gadgetHandler:IsSyncedCode() then
...
+		local currentCommand = commands and commands[1]
+		if
+			not currentCommand
+			or currentCommand.id ~= CMD_ATTACK
+			or not currentCommand.params
+			or #currentCommand.params < 3
+			or not currentCommand.options
+			or not currentCommand.options.internal
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: I'd assume these do not need nil checks.

File: luarules/gadgets/unit_areaattack.lua [hunk 9]
```diff
@@ -108,15 +142,71 @@ if gadgetHandler:IsSyncedCode() then
...
+			return
+		end
+
+		activeAttacks[ownerID] = nil
+		-- Generated ground attacks are persistent, so finish this shot after one
+		-- complete salvo and let the area command choose another random position.
+		if commands[2] then
+			finishedAttacks[ownerID] = attack
+		end
+	end
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: We know the duration of a salvo. Why not just schedule a watch frame and poll on it?

**Review by efrec — COMMENTED** [review]
Seems to fix the issue but works hard doing it. See notes.

**Comment by efrec** [comment]
Looks like I missed a lot of conversation. The proposed behavior change does scan close to a manual fire -- a single volley is an atomic command that acts like an ability, thus should control similarly to e.g. Thor launcher, like Chronographer mentioned.

**Comment by eun-ice** [comment]
> Seems to fix the issue but works hard doing it. See notes.

Yes. The best way would be a burstend callin from the engine. 

Proposed in https://github.com/beyond-all-reason/RecoilEngine/pull/3286

Until that is merged I proposed this weird workaround that I don't really like either. 

But I think I can address all your concerns. Maybe we don't have to wait for the engine.

**Comment by eun-ice** [comment]
> Looks like I missed a lot of conversation. The proposed behavior change does scan close to a manual fire -- a single volley is an atomic command that acts like an ability, thus should control similarly to e.g. Thor launcher, like Chronographer mentioned.

I moved that to a separate PR. 

Currently I am working on cataloging all the different questions that came up during my work on attack/attack ground and set target and how they interact. 

Maybe that will help us make get a decision on where current game behavior is intended and where we want it changed, instead of debating this case by case.

Until then I will only open clear bugfixes as PR.

File: luarules/gadgets/unit_areaattack.lua (hunk 3 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: > An alias should identify its contents (that's the point), e.g. areaWeaponRange, areaWeaponSalvo.

done.

File: luarules/gadgets/unit_areaattack.lua (hunk 5 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: done

File: luarules/gadgets/unit_areaattack.lua (hunk 6 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua:158]: Correct.

I have pushed another idea for a workaround by querying  Spring.GetUnitWeaponState(unitID, 1, "salvoLeft") and then Spring.GetUnitWeaponState(unitID, 1, "nextSalvo") to determine when to recheck.

In my tests with a local engine that seemed to work too.

This might even be more correct, if not all projectiles are shot in a salvo, e.g. due to obstructions.

File: luarules/gadgets/unit_areaattack.lua (hunk 9 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: yes, that idea did lead to a cleaner and probably even more correct solution.

File: luarules/gadgets/unit_areaattack.lua (hunk 8 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Emmy disagreed. I did that just to make Emmy happy. Emmy doesn't seem to like me, let's see if Emmy is ok now.

File: luarules/gadgets/unit_areaattack.lua (hunk 7 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Good catch. This was a leftover from the earlier behavior change push, where we needed to check if there was a persistent attack command after.
Is fixed.

File: luarules/gadgets/unit_areaattack.lua (hunk 4 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: Done.

**Comment by eun-ice** [comment]
I updated the solution to be more efficient, instead of counting projectiles we now ask the engine for Spring.GetUnitWeaponState(unitID, 1, "salvoLeft") and then wait until Spring.GetUnitWeaponState(unitID, 1, "nextSalvo")

However, there are instances where the unit might not be able to shoot at the randomly chosen point, in that case nothing would happen and the unit would not choose a new point. Related Bug: https://github.com/beyond-all-reason/RecoilEngine/pull/3241

But that is a rare case and not a new bug, so maybe something for the big "do later" list?

File: luarules/gadgets/unit_areaattack.lua [hunk 10]
```diff
@@ -19,20 +19,32 @@ local CMD_AREA_ATTACK_GROUND = GameCMD.AREA_ATTACK_GROUND
...
 
 	local canAreaAttack = {}
+	local areaAttackWeaponDefs = {}
+	local areaAttackWeaponDefByUnitDef = {}
 	for unitDefID, unitDef in pairs(UnitDefs) do
 		if #unitDef.weapons > 0 and unitDef.customParams.canareaattack then
-			canAreaAttack[unitDefID] = WeaponDefs[unitDef.weapons[1].weaponDef].range
+			local weaponDefID = unitDef.weapons[1].weaponDef
+			local weaponDef = WeaponDefs[weaponDefID]
+			if weaponDef then
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: Cannot be nil

File: luarules/gadgets/unit_areaattack.lua [hunk 11]
```diff
@@ -52,12 +64,14 @@ if gadgetHandler:IsSyncedCode() then
...
-				Spring.GiveOrderToUnit(
-					o.unit,
-					CMD.INSERT,
-					{ 0, CMD.ATTACK, 0, o.x + math_cos(phase) * amp, o.y, o.z + math_sin(phase) * amp },
-					{ "alt" }
-				)
+				Spring.GiveOrderToUnit(o.unit, CMD.INSERT, {
+					0,
+					CMD.ATTACK,
+					CMD.OPT_INTERNAL,
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: Prefer the local

File: luarules/gadgets/unit_areaattack.lua [hunk 12]
```diff
@@ -108,15 +153,47 @@ if gadgetHandler:IsSyncedCode() then
...
+		local unitDefID = Spring.GetUnitDefID(ownerID)
+		if areaAttackWeaponDefByUnitDef[unitDefID] ~= weaponDefID then
+			return
+		end
+
+		local commandID, commandOptions, commandTag, _, _, targetZ = Spring.GetUnitCurrentCommand(ownerID)
+		if commandID ~= CMD_ATTACK or targetZ == nil or math_bit_and(commandOptions, CMD_OPT_INTERNAL) == 0 then
+			return
+		end
+		if not Spring.GetUnitCurrentCommand(ownerID, 2) then
```
Comment (efrec) [inline luarules/gadgets/unit_areaattack.lua]: Why not check that this is == our command here?

**Review by efrec — CHANGES_REQUESTED** [review]

File: luarules/gadgets/unit_areaattack.lua (hunk 10 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: > Cannot be nil

I know, but Emmy, Emmy came after me, Emmy is scary.

File: luarules/gadgets/unit_areaattack.lua (hunk 11 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: done.

File: luarules/gadgets/unit_areaattack.lua (hunk 12 above)
Comment (eun-ice) [reply luarules/gadgets/unit_areaattack.lua]: This was wrong entirely. Fixed now.

**Review by efrec — APPROVED** [review]
